### PR TITLE
implementing keras integration

### PR DESCRIPTION
**Pull request by higumachan**
I implemented `KerasPruningCallback` like `ChainerPruningExtension`.

This can be used as follows.

```python3
    def objective(trial):
        # type: (optuna.trial.Trial) -> float

        model = Sequential()
        model.add(Dense(1, activation='sigmoid', input_dim=20))
        model.compile(optimizer='rmsprop',
                      loss='binary_crossentropy',
                      metrics=['accuracy'])
        X = np.zeros((16, 20), np.float32)
        y = np.zeros((16,), np.int32)
        model.fit(
            X,
            y,
            batch_size=1,
            epochs=1,
            callbacks=[KerasPruningCallback(trial, 'acc')],
            verbose=0
        )

        return model.evaluate(X, y)
```

### CODE REVIEW

**Comment by higumachan** [comment]
tensorflow can not install on python3.7 with pip.

I think that there are two possible solutions.

1. skip test and raise Error using `KerasPruningCallback` with python3.7.
2. tensorflow install from wheel file.

Does anyone have a good solution?

File: optuna/integration/keras.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    Args:
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of the
+            objective function.
+        observation_key:
+            An evaluation metric for pruning, e.g., ``val/loss`` and
+            ``val/acc``. Please refer to
+            `keras.Model reference <https://keras.io/models/about-keras-models/>`_ for further details. 
+    """  # NOQA
```
Comment (sile) [inline optuna/integration/keras.py]: I think this `NOQA` can be removed if you insert an appropriate newline to the above line.

And please execute the `$ autopep8 optuna/integration/keras.py --in-place` command to fix some style problems that cannot be detected by `flake8`.

File: optuna/integration/keras.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            raise optuna.structs.TrialPruned(message)
+
+
+def _check_keras_availability():
+    # type: () -> None
+
+    if not _available:
+        raise ImportError(
+            'Keras is not available. Please install Keras to use this feature. '
+            'Keras can be installed by executing `$ pip install chainer`. '
```
Comment (sile) [inline optuna/integration/keras.py]: typo: "chainer"

File: tests/integration_tests/test_keras.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    study.optimize(objective, n_trials=1)
+    assert study.trials[0].state == optuna.structs.TrialState.PRUNED
+
+    study = optuna.create_study(pruner=DeterministicPruner(False))
+    study.optimize(objective, n_trials=1)
+    assert study.trials[0].state == optuna.structs.TrialState.COMPLETE
+    assert study.trials[0].value == 1.0
+
+
+def test_chainer_pruning_extension_observation_isnan():
```
Comment (sile) [inline tests/integration_tests/test_keras.py]: typo: "chainer"

**Comment by higumachan** [comment]
Thank you for reviewing, I fixed that review points.

**Comment by toshihikoyanase** [comment]
> tensorflow can not install on python3.7 with pip.
> 
> I think that there are two possible solutions.
> 
> 1. skip test and raise Error using `KerasPruningCallback` with python3.7.
> 2. tensorflow install from wheel file.
> 
> Does anyone have a good solution?

Thank you for your contribution.
I think both choices have pros and cons.

In my opinion, we can just skip `test_keras.py` in Python 3.7 environment. This is partly because Tensorflow does not support Python 3.7 currently, and partly because some skillful users can use Tensorflow with Python 3.7 by using some hacks as you mentioned in the second choice.

So, Optuna does not provide official support of `KerasPruningCallback` in Python 3.7 but provides the functionality itself for advanced users.

I checked the way to skip `test_keras.py` in the `tests-python37` job of CircleCI.

1.  Update `setup.py`. In [`extras_require` of `get_extras_require` function](https://github.com/pfnet/optuna/blob/master/setup.py#L26), please extract `keras, tensorflow` to the new entry `"keras"` as follows:

```
        'testing': ['pytest', 'mock', 'bokeh', 'chainer>=5.0.0', 'xgboost',
                    'mpi4py', 'lightgbm'],
        'keras': ['keras', 'tensorflow'],
```

2. Update `.circleci/config.yml`. Please add `install-keras` step to `tests-python36` to install keras as follows: 

```
  tests-python36:
    docker:
      - image: circleci/python:3.6
    steps:
      - checkout
      - run: *install

      - run: &install-keras
          name: install-keras
          command: |
            . venv/bin/activate
            pip install .[keras]

      - run: *tests
      - run: *tests-mn
```

3. Add references to `install-keras` to `tests-python35`, `tests-python34`, `tests-python27`, `examples-python36`, `examples-python35`, `examples-python34`, and `examples-python27`.

```
    steps:
      [checkout, run: *install, run: *install-keras, run: *tests, run: *tests-mn]
```

4. Skip `tests_keras.py` if `keras` is not installed as follows:

```
try:
    from keras.layers import Dense
    from keras import Sequential
    _available = True
except ImportError:
    _available = False

def test_keras_pruning_callback():
    # type: () -> None

    if not _available:
        pytest.skip('This test requires keras.')
```

5. Done.

This is just an example, so please improve it if you come up with better implementation.
Also, please add `TODO` comments to remove this fix after Tensorflow supports Python 3.7 officially.

@iwiwi @g-votte @sile 
If you have comments, please let us know.

**Comment by higumachan** [comment]
@toshihikoyanase

Thank you for giving me good way.

I will modification it based on your solution!

**Comment by higumachan** [comment]
@sile Thank you for reviewing, I fixed that points.

@toshihikoyanase 
I modified testing refer to your example.

File: optuna/integration/keras.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                callbacks=KerasPruningCallback(trial, 'val/loss'))
+
+    Args:
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of the
+            objective function.
+        observation_key:
+            An evaluation metric for pruning, e.g., ``val/loss`` and
+            ``val/acc``. Please refer to `keras.Model reference
+            <https://keras.io/models/about-keras-models/>`_ for further details.
```
[the remark this answers is not in the document]
Comment (higumachan) [reply optuna/integration/keras.py]: @toshihikoyanase 

Thank you for your review!

I will change that.

```
monitor:
             An evaluation metric for pruning, e.g., ``val_loss`` or
            ``val_acc``. Please refer to `keras.Callback reference
            <https://keras.io/callbacks/#callback>`_ for further details.
```

Is it good enough?

**Comment by higumachan** [comment]
@toshihikoyanase 
Thank you for reviewing, I will fix that points.

File: optuna/integration/keras.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (toshihikoyanase) [reply optuna/integration/keras.py]: Looks good to me. Thanks!

**Comment by higumachan** [comment]
@toshihikoyanase 
I fixed your review points thank you.

File: optuna/integration/keras.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+class KerasPruningCallback(Callback):
+    """Keras callback to prune unpromising trials.
+
+    Example:
+
+        Add a pruning callback which observes validation losses.
+
+        .. code::
+
+            model.fit(X, y, callbacks=KerasPruningCallback(trial, 'val/loss'))
```
Comment (toshihikoyanase) [inline optuna/integration/keras.py]: I think `val/loss` is a typo for `val_loss` as we discussed in the previous review.

Comment (higumachan) [reply optuna/integration/keras.py]: Oops, It is a typo. I will fix it ASAP.
